File: relay/crypto.js
/*
 * vim: ts=4:sw=4:expandtab
 */

'use strict';

const libsignal = require('libsignal');


module.exports = {

    // Decrypts message into a raw string
    decryptWebsocketMessage: function(message, signaling_key) {
        if (signaling_key.byteLength != 52) {
            throw new Error("Got invalid length signaling_key");
        }
        if (message.byteLength < 1 + 16 + 10) {
            throw new Error("Got invalid length message");
        }
        if (message[0] != 1) {
            throw new Error("Got bad version number: " + message[0]);
        }
        var aes_key = signaling_key.slice(0, 32);
        var mac_key = signaling_key.slice(32, 32 + 20);
        var iv = message.slice(1, 17);
        var ciphertext = message.slice(1 + 16, message.byteLength - 10);
        var ivAndCiphertext = message.slice(0, message.byteLength - 10);
        var mac = message.slice(message.byteLength - 10, message.byteLength);
        libsignal.crypto.verifyMAC(ivAndCiphertext, mac_key, mac, 10);
        return libsignal.crypto.decrypt(aes_key, ciphertext, iv);
    },

    decryptAttachment: function(encryptedBin, keys) {
        if (keys.byteLength != 64) {
            throw new Error("Got invalid length attachment keys");
        }
        if (encryptedBin.byteLength < 16 + 32) {
            throw new Error("Got invalid length attachment");
        }
        var aes_key = keys.slice(0, 32);
        var mac_key = keys.slice(32, 64);
        var iv = encryptedBin.slice(0, 16);
        var ciphertext = encryptedBin.slice(16, encryptedBin.byteLength - 32);
        var ivAndCiphertext = encryptedBin.slice(0, encryptedBin.byteLength - 32);
        var mac = encryptedBin.slice(encryptedBin.byteLength - 32, encryptedBin.byteLength);
        libsignal.crypto.verifyMAC(ivAndCiphertext, mac_key, mac, 32);
        return libsignal.crypto.decrypt(aes_key, ciphertext, iv);
    },

    encryptAttachment: function(plaintext, keys, iv) {
        if (keys.byteLength != 64) {
            throw new Error("Got invalid length attachment keys");
        }
        if (iv.byteLength != 16) {
            throw new Error("Got invalid length attachment iv");
        }
        const aes_key = keys.slice(0, 32);
        const mac_key = keys.slice(32, 64);
        const ciphertext = libsignal.crypto.encrypt(aes_key, plaintext, iv);
        const ivAndCiphertext = Buffer.concat([iv, ciphertext]);
        const mac = libsignal.crypto.sign(mac_key, ivAndCiphertext);
        return Buffer.concat([ivAndCiphertext, mac]);
    }
};

File: librelay/crypto.py
import hashlib
import hmac
import logging
from . import executor
from libsignal.sessioncipher import AESCipher

logger = logging.getLogger(__name__)


async def sign(*args):
    return await executor.run(_sign, *args)


def _sign(key, data):
    m = hmac.new(key, digestmod=hashlib.sha256)
    m.update(data)
    return m.digest()


async def verifyMAC(*args):
    return await executor.run(_verifyMAC, *args)


def _verifyMAC(data, key, mac):
    calculated_mac = _sign(key, data)
    length = len(mac)
    if mac != calculated_mac[:length]:
        raise Exception("Bad MAC")


async def decryptWebSocketMessage(*args):
    return await executor.run(_decryptWebSocketMessage, *args)


def _decryptWebSocketMessage(message, signaling_key):
    """ Decrypts message into a raw string """
    if len(signaling_key) != 52:
        raise ValueError("Got invalid length signaling_key")
    if len(message) < 1 + 16 + 10:
        raise ValueError("Got invalid length message")
    if message[0] != 1:
        raise ValueError("Got bad version number: " + message[0])
    aes_key = signaling_key[:32]
    mac_key = signaling_key[32:32 + 20]
    iv = message[1:17]
    ciphertext = message[1 + 16:-10]
    ivAndCiphertext = message[:-10]
    mac = message[-10:]
    _verifyMAC(ivAndCiphertext, mac_key, mac)
    return AESCipher(aes_key, iv).decrypt(ciphertext)


async def decryptAttachment(*args):
    return await executor.run(_decryptAttachment, *args)


def _decryptAttachment(encryptedBin, keys):
    if len(keys) != 64:
        raise ValueError("Got invalid length attachment keys")
    if len(encryptedBin) < 16 + 32:
        raise ValueError("Got invalid length attachment")
    aes_key = keys[:32]
    mac_key = keys[32:64]
    iv = encryptedBin[:16]
    ciphertext = encryptedBin[16:-32]
    ivAndCiphertext = encryptedBin[:-32]
    mac = encryptedBin[-32:]
    _verifyMAC(ivAndCiphertext, mac_key, mac)
    return AESCipher(aes_key, iv).decrypt(ciphertext)


async def encryptAttachment(*args):
    return await executor.run(_encryptAttachment, *args)


def _encryptAttachment(plaintext, keys, iv):
    if len(keys) != 64:
        raise ValueError("Got invalid length attachment keys")
    if len(iv) != 16:
        raise ValueError("Got invalid length attachment iv")
    aes_key = keys[:32]
    mac_key = keys[32:64]
    ciphertext = AESCipher(aes_key, iv).encrypt(plaintext)
    ivAndCiphertext = iv + ciphertext
    mac = _sign(mac_key, ivAndCiphertext)
    return ivAndCiphertext + mac
